app/assets/javascripts/actions/revisions_actions.js

```diff
@@ -9,9 +11,349 @@ import {
+  const allData = [];
+  let continueToken;
+  let hasMore = true;
+  while (hasMore) {
+    let response;
+    if (continueToken) {
+      params[continue_str] = continueToken[continue_str];
+      params.continue = continueToken.continue;
+    }
+    try {
+      response = await request(`${API_URL}?${stringify(params)}&origin=*`);
+      if (!response.ok) {
+        throw response;
+      }
+    } catch (e) {
+      return allData;
+    }
+    const json = await response.json();
+    allData.push(...json.query.usercontribs);
+    if (json.continue) {
+      continueToken = json.continue;
+    } else {
+      hasMore = false;
+    }
+  }
+  return allData;
+};
+const fetchAllRevisions = async (API_URL, days, usernames, wiki, course_start, last_date) => {
+  let ucend;
+  if (moment(last_date).subtract(days, 'days').isBefore(course_start)) {
+    ucend = moment(course_start).format();
+  } else {
+    ucend = moment(last_date).subtract(days, 'days').format();
+  }
+  // eslint-disable-next-line no-console
+  console.log(`Fetching revisions between ${moment(last_date).format()} and ${ucend} -> ${days} days from ${API_URL}`);
+  // since a max of 50 users are allowed in one query
+  const usernamesChunks = chunk(usernames, 50);
+  const usernamePromises = [];
+  /* eslint-disable no-restricted-syntax */
+  for (const usernameChunk of usernamesChunks) {
+    const params = {
+      action: 'query',
+      format: 'json',
+      list: 'usercontribs',
+      ucuser: usernameChunk.join('|'),
+      ucprop: 'ids|title|sizediff|timestamp',
+      uclimit: 50,
+      ucend,
+      ucstart: moment(last_date).format(),
+      ucdir: 'older',
+    };
+    usernamePromises.push(fetchAll(API_URL, params, 'uccontinue'));
+  }
+  const values = await Promise.all(usernamePromises);
+  const revisions = flatten(values);
+  return revisions;
+};
+
+const fetchClassFromRevisionsOfWiki = async (wiki_url, revisionsOfWiki) => {
+  // remove duplicates -> each article occurs only once after this
+  const prefix = `https://${wiki_url}`;
+  const API_URL = `${prefix}/w/api.php`;
+  // eslint-disable-next-line no-console
+  console.log(`Fetching page assessments from ${wiki_url}`);
+
+  const uniqueArticles = [...new Set(revisionsOfWiki.filter(revision => revision.ns === 0).map(revision => revision.title))];
+  const allRatings = [];
+  const chunks = chunk(uniqueArticles, 50);
+  for (const uniqueArticlesChunk of chunks) {
+    const params = {
+      action: 'query',
+      format: 'json',
+      titles: uniqueArticlesChunk.join('|'),
+      prop: 'pageassessments',
+      pasubprojects: false,
+      palimit: 500
+    };
+    const response = await request(`${API_URL}?${stringify(params)}&origin=*`);
+    const pages = (await response.json())?.query?.pages;
+    if (pages) { allRatings.push(pages); }
+  }
+  if (!allRatings) {
+    // no ratings found
+    return;
+  }
+  // merge list of objects of ratings into a single object
```

Comment:
I think all this processing of the results should be extracted into utility functions, with a clear explanation in comments of what shape of data it expects and what shape of data it outputs.